string.c

```diff
@@ -240,6 +244,11 @@ rb_str_size_as_embedded(VALUE str)
     else {
         real_size = sizeof(struct RString);
     }
+
+    if (FL_TEST_RAW(str, STR_PRECOMPUTED_HASH)) {
+        real_size += sizeof(st_index_t);
+    }
```

Comment:
This was the bug. The compactor thought the string could fit in a smaller slot, corrupting the precomputed hash.